irs/core/upload_handler.py

```diff
@@ -0,0 +1,235 @@
+# Copyright 2021 - 2022 Universität Tübingen, DKFZ and EMBL
+# for the German Human Genome-Phenome Archive (GHGA)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Core interrogation room logic called by event subscriber"""
+
+import base64
+import codecs
+import hashlib
+import math
+from typing import List, Sequence, Tuple
+
+import requests
+from crypt4gh.lib import CIPHER_SEGMENT_SIZE, decrypt_block
+from hexkit.providers.akafka import KafkaEventPublisher
+from hexkit.providers.s3.provider import S3Config, S3ObjectStorage
+
+from irs.adapters.outbound.kafka_producer import (
+    UploadValidationFailureEvent,
+    UploadValidationSuccessEvent,
+)
+from irs.config import CONFIG
+from irs.core import exceptions
+from irs.core.http_translation import ResponseExceptionTranslator
+
+PART_SIZE = 16 * 1024**2
+
+
+async def process_new_upload(  # pylint: disable=too-many-locals
+    *, object_id: str, object_size: int, public_key: str, checksum: str
+):
+    """
+    Forwards first file part to encryption key store, retrieves file encryption
+    secret(s) (K_data), decrypts file and computes checksums
+    """
+    download_url = await get_download_url(object_id=object_id)
+    part = await retrieve_part(url=download_url, start=0, stop=PART_SIZE - 1)
+
+    secret, secret_id, offset = await send_to_eks(
+        file_part=part, public_key=public_key, eks_url=CONFIG.eks_url
+    )
+    part_checksums, total_checksum = await compute_checksums(
+        download_url=download_url,
+        secret=secret,
+        object_size=object_size,
+        offset=offset,
+    )
+
+    type_ = "irs_publisher"
+    key = "irs"
+    async with KafkaEventPublisher.construct(config=CONFIG) as publisher:
+        if total_checksum == checksum:
+            topic = "upload_validation_success"
+            data = UploadValidationSuccessEvent(
+                content_id=total_checksum,
+                secret_id=secret_id,
+                offset=offset,
+                part_size=PART_SIZE,
+                part_checksums_md5=part_checksums[0],
+                part_checksums_sha256=part_checksums[1],
+            ).dict()
+        else:
+            topic = "upload_validation_failure"
```

Comment:
See above.

Reply:
We can use the same topic for all events published by this service. We might call it by default something like file_interrogations.